### ISSUE TITLE
Version 1.11 error with Nvidia ShieldTV 2015 (Android 9 Pie / SDK Version 28)

### ISSUE THREAD

**Issue by tam1m** [Genymobile/scrcpy#940]
After updating from `1.10` to `1.11` scrcpy is no longer working with my ShieldTV.

```
➜ scrcpy
INFO: scrcpy 1.11 <https://github.com/Genymobile/scrcpy>
/usr/local/share/scrcpy/scrcpy-server: 1 file pushed. 5.8 MB/s (25454 bytes in 0.004s)
[server] ERROR: Exception on thread Thread[main,5,main]
java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.Context android.app.Application.createPackageContext(java.lang.String, int)' on a null object reference
at android.media.MediaCodec.isSupported(MediaCodec.java:1794)
at android.media.MediaCodec.checkAppForAutoFrc(MediaCodec.java:1831)
at android.media.MediaCodec.configureCodecParams(MediaCodec.java:1980)
at android.media.MediaCodec.<init>(MediaCodec.java:2011)
at android.media.MediaCodec.createEncoderByType(MediaCodec.java:1944)
at com.genymobile.scrcpy.ScreenEncoder.createCodec(ScreenEncoder.java:138)
at com.genymobile.scrcpy.ScreenEncoder.streamScreen(ScreenEncoder.java:63)
at com.genymobile.scrcpy.Server.scrcpy(Server.java:32)
at com.genymobile.scrcpy.Server.main(Server.java:146)
at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:340)
INFO: Initial texture: 1920x1080
WARN: Device disconnected
```

**Comment by rom1v** [comment]
Thank you for your report.

If you try again v1.10, does it work?

**Comment by tam1m** [comment]
Yes, v1.10 works perfectly fine

**Comment by rom1v** [comment]
Could you retry with this change, please?

```diff
diff --git a/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java b/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java
index c9a37f8..504e954 100644
--- a/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java
+++ b/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java
@@ -53,7 +53,6 @@ public class ScreenEncoder implements Device.RotationListener {
 
     public void streamScreen(Device device, FileDescriptor fd) throws IOException {
         Workarounds.prepareMainLooper();
-        Workarounds.fillAppInfo();
 
         MediaFormat format = createFormat(bitRate, maxFps, iFrameInterval);
         device.setRotationListener(this);
```

**Comment by tam1m** [comment]
That's it! It works with your patch applied.

**Comment by rom1v** [comment]
OK, so this conflicts with #365 (which works only with the patch). :disappointed:

**Comment by tam1m** [comment]
Would it be sensible to call it only for devices that are known to not work without it?
Something like this
```
if(needs_workaround_device_list.contains(Build.PRODUCT))
  Workarounds.fillAppInfo();
```

**Comment by rom1v** [comment]
If you restore `Workarounds.fillAppInfo()` and apply this patch instead, what is the result?

```diff
diff --git a/server/src/main/java/com/genymobile/scrcpy/Workarounds.java b/server/src/main/java/com/genymobile/scrcpy/Workarounds.java
index f45d82a..ba5cd21 100644
--- a/server/src/main/java/com/genymobile/scrcpy/Workarounds.java
+++ b/server/src/main/java/com/genymobile/scrcpy/Workarounds.java
@@ -1,11 +1,15 @@
 package com.genymobile.scrcpy;
 
 import android.annotation.SuppressLint;
+import android.app.Application;
+import android.app.Instrumentation;
+import android.content.Context;
 import android.content.pm.ApplicationInfo;
 import android.os.Looper;
 
 import java.lang.reflect.Constructor;
 import java.lang.reflect.Field;
+import java.lang.reflect.Method;
 
 public final class Workarounds {
     private Workarounds() {
@@ -56,6 +60,14 @@ public final class Workarounds {
             Field mBoundApplicationField = activityThreadClass.getDeclaredField("mBoundApplication");
             mBoundApplicationField.setAccessible(true);
             mBoundApplicationField.set(activityThread, appBindData);
+
+            Method getSystemContextMethod = activityThreadClass.getDeclaredMethod("getSystemContext");
+            // this is not a full Android context giving access to everything
+            Context ctx = (Context) getSystemContextMethod.invoke(activityThread);
+            Application app = Instrumentation.newApplication(Application.class, ctx);
+            Field mInitialApplicationField = activityThreadClass.getDeclaredField("mInitialApplication");
+            mInitialApplicationField.setAccessible(true);
+            mInitialApplicationField.set(activityThread, app);
         } catch (Throwable throwable) {
             // this is a workaround, so failing is not an error
             Ln.w("Could not fill app info: " + throwable.getMessage());
```

**Comment by tam1m** [comment]
Just tried and it works

**Comment by rom1v** [comment]
Merged on `dev`: 2b845680b0d8d6778600d3c46a570d73dc83e2b5

**Comment by AceHack** [comment]
Is there somewhere I can download dev builds?

**Comment by rom1v** [comment]
No, this will be in v1.12 (which should be released soon, hopefully).

Meanwhile, you can use [v1.10](https://github.com/Genymobile/scrcpy/releases/tag/v1.10).

**Comment by rom1v** [comment]
(v1.12 which fixes the problem is out)

**Comment by rom1v** [comment]
To fix #994, I changed the way the workaround is called.

Could you please take scrcpy v1.13 and replace this file:

[`scrcpy-server`](https://tmp.rom1v.com/scrcpy/issue994/1/scrcpy-server)
_SHA256: b64e6672eeb2b7c00c6f83673783bd8839ba65ea2ac7edcdeeca44676047328b_

and confirm that it works without the error message.

Thank you.

**Comment by rom1v** [comment]
Could you please test #5154 and confirm that it still works?
